src/metric-engine/src/engine/alter.rs

```diff
@@ -12,148 +12,157 @@
 use store_api::region_request::{AffectedRows, AlterKind, RegionAlterRequest};
 use store_api::storage::RegionId;
+use validate::validate_alter_region_requests;
 
 use crate::engine::MetricEngineInner;
 use crate::error::{
     LogicalRegionNotFoundSnafu, PhysicalRegionNotFoundSnafu, Result, SerializeColumnMetadataSnafu,
+    UnexpectedRequestSnafu,
 };
-use crate::utils::{to_data_region_id, to_metadata_region_id};
+use crate::utils::to_data_region_id;
 
 impl MetricEngineInner {
-    /// Dispatch region alter request
-    pub async fn alter_region(
+    pub async fn alter_regions(
         &self,
-        region_id: RegionId,
-        request: RegionAlterRequest,
+        mut requests: Vec<(RegionId, RegionAlterRequest)>,
         extension_return_value: &mut HashMap<String, Vec<u8>>,
     ) -> Result<AffectedRows> {
-        let is_altering_physical_region = self.is_physical_region(region_id);
+        if requests.is_empty() {
+            return Ok(0);
+        }
 
-        let result = if is_altering_physical_region {
-            self.alter_physical_region(region_id, request).await
+        let first_region_id = &requests.first().unwrap().0;
+        if self.is_physical_region(*first_region_id) {
+            ensure!(
+                requests.len() == 1,
+                UnexpectedRequestSnafu {
+                    reason: "Physical table must be altered with single request".to_string(),
+                }
+            );
+            let (region_id, request) = requests.pop().unwrap();
+            self.alter_physical_region(region_id, request).await?;
         } else {
-            let physical_region_id = self.alter_logical_region(region_id, request).await?;
-
-            // Add physical table's column to extension map.
-            // It's ok to overwrite existing key, as the latter come schema is more up-to-date
-            let physical_columns = self
-                .data_region
-                .physical_columns(physical_region_id)
+            self.alter_logical_regions(requests, extension_return_value)
                 .await?;
-            extension_return_value.insert(
-                ALTER_PHYSICAL_EXTENSION_KEY.to_string(),
-                ColumnMetadata::encode_list(&physical_columns)
-                    .context(SerializeColumnMetadataSnafu)?,
-            );
-
-            Ok(())
-        };
-
-        result.map(|_| 0)
+        }
+        Ok(0)
     }
 
-    /// Return the physical region id behind this logical region
-    async fn alter_logical_region(
+    /// Alter multiple logical regions on the same physical region.
+    pub async fn alter_logical_regions(
         &self,
-        logical_region_id: RegionId,
-        request: RegionAlterRequest,
-    ) -> Result<RegionId> {
+        requests: Vec<(RegionId, RegionAlterRequest)>,
+        extension_return_value: &mut HashMap<String, Vec<u8>>,
+    ) -> Result<AffectedRows> {
+        validate_alter_region_requests(&requests)?;
```

Comment:
if alter region in batches only support add columns should we rename some methods to emphasize this?